### PR TITLE
CSHARP-2003: $bitsAnySet etc should work with Enums.

### PR DESCRIPTION
**Pull request by rstam**

### CODE REVIEW

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/BitMaskComparisonExpressionToFilterTranslator.cs [hunk 1]
```diff
@@ -24,11 +24,49 @@ namespace MongoDB.Driver.Linq.Linq3Implementation.Translators.ExpressionToFilter
...
-        public static bool CanTranslate(Expression leftExpression)
+        public static bool CanTranslate(Expression leftExpression, Expression rightExpression)
         {
-            return
-                leftExpression is BinaryExpression leftBinaryExpression &&
-                leftBinaryExpression.NodeType == ExpressionType.And;
+            return CanTranslate(leftExpression, rightExpression, out _);
+        }
+
+        public static bool CanTranslate(Expression leftExpression, Expression rightExpression, out BinaryExpression leftBinaryExpression)
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/BitMaskComparisonExpressionToFilterTranslator.cs:32]: There are two patterns that this translator can translate:

x.E & mask == value
Convert(Convert((Convert(x.E, Int32) & mask), E), Int32) == value

The second is for enums where Int32 is the underlying type.

Both forms are translated the same way, we just have to deal with the `Convert`s

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/BitMaskComparisonExpressionToFilterTranslator.cs [hunk 2]
```diff
@@ -39,10 +77,9 @@ public static bool CanTranslate(Expression leftExpression)
             AstComparisonFilterOperator comparisonOperator,
             Expression rightExpression)
         {
-            if (leftExpression is BinaryExpression leftBinaryExpression &&
-                leftBinaryExpression.NodeType == ExpressionType.And)
+            if (CanTranslate(leftExpression, rightExpression, out var leftBinaryExpression))
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/BitMaskComparisonExpressionToFilterTranslator.cs:80]: Leverage `CanTranslate` to give us the inner `leftBinaryExpression` whether or not an Enum is involved.

File: src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/BitMaskComparisonExpressionToFilterTranslator.cs [hunk 3]
```diff
@@ -39,10 +77,9 @@ public static bool CanTranslate(Expression leftExpression)
...
             {
-                var fieldExpression = leftBinaryExpression.Left;
+                var fieldExpression = ConvertHelper.RemoveConvertToEnumUnderlyingType(leftBinaryExpression.Left);
```
Comment (rstam) [inline src/MongoDB.Driver/Linq/Linq3Implementation/Translators/ExpressionToFilterTranslators/ExpressionTranslators/BitMaskComparisonExpressionToFilterTranslator.cs:82]: This is a no-op if `Left` is not an enum convert.

File: tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp2003Tests.cs [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        [Flags]
+        private enum E
+        {
+            E1 = 1,
+            E2 = 2,
+            E4 = 4,
+            E8 = 8
+        }
+
+        private class C
```
Comment (rstam) [inline tests/MongoDB.Driver.Tests/Linq/Linq3ImplementationTests/Jira/CSharp2003Tests.cs:163]: I renamed the class and the enum and the members to be more self-describing.

**Review by JamesKovacs — APPROVED** [review]
LGTM

**Pull request merged**